text/3724-theupdateframework.md

```diff
@@ -0,0 +1,341 @@
+
+The API surface of this crate is to be determined upon implementation in Cargo and Rustup. However, because both tools will need to perform synchronization and validation against the tuf-root repository, they shall used this shared interface to guarantee compatibility.
+
+This API will include operations to sync the TUF repositories efficiently, and to perform a verified download of an object.
+
+## TAP-16 Implementation
+
+We're proposing to use [TAP-16](https://github.com/theupdateframework/taps/blob/master/tap16.md) to provide efficient update checking and download sizes. TAP-16 uses Merkle trees rather than full lists for the download of a snapshot of the inventory of a repository (`snapshot.json`). We want to ensure that, as crates.io grows, the total size clients have to download when checking for updates remains small.
+
+### Shared local folder changes
+
+Creation of a new `~/.cargo/tuf` directory. (If Cargo stores its registry information in another directory, the `tuf` directory should be stored alongside the `registry` directory.) This directory shall be used for all TUF operations by project tools (both Rustup and Cargo). The cargo folder was chosen as the main location of residence for these files given that although Rustup will be performing the initialization of these folders, there is already a precedent set for shared files living within the cargo folder.
+
+- `~/.cargo/tuf` The top-level directory of local copies of TUF repositories
+- `~/.cargo/tuf/root` a copy of the `tuf-root` repository locally synchronized
+- `~/.cargo/tuf/crates` a copy of the `tuf-crates` repository locally synchronized
+
+### Controlling enablement of TUF
+
+This RFC does not specify how to handle non-`crates.io` repositories. Cargo can choose to enable TUF for third-party repositories in the future, or may default to only using TUF for crates.io unless otherwise configured. Cargo might choose to use an environment variable (e.g. `CARGO_TUF_DISABLE`) to disable all usage of TUF on third-party repositories.
+
+## Cargo changes
+- Addition of `cargo-tuf-lib` as a dependency
+- `cargo-tuf-lib::sync` attempted prior to an index update
+- `cargo-tuf-lib::verify_snapshot` called on an index update on the entire index
+
+## Rustup Changes
+
+- Addition of `cargo-tuf-lib` as a dependency
+- `cargo-tuf-lib::sync` attempted on execution
+- `cargo-tuf-lib::verify_snapshot` called on an index update on the entire index
+- `cargo-tuf-lib::verify_file` called on the download of any release files (stable and nightly), or self-updates
+
+## Crates.io changes
+
+- Prior to updating the index, crates.io shall perform the online signing of the index entry to update the targets and sign the index entry, saving this as a like-pathed artifact in the TUF repository.
```

Comment:
from just this sentence alone I unfortunately have no clue what that actually means for us 😅 

side note: what happens if the update of the TUF repository is successful but then the index repo update fails? or if the sparse index upload fails? what happens if both indexes are out-of-sync?

the indexes are treated as eventually consistent by crates.io. I'm not sure if we can actually guarantee a TUF repo update happening before and close in time to the two separate index updates.

Reply:
Yes, I think implementation wise the TUF copy of the index is going to have to be a 3ed eventually consistent source. At least at first.

At some point we could start maintaining the Git index using an entirely separate server that watches for changes in the TUF index then downloads them then commits them. Similarly we could have a separate service that watches the TUF index and uploads the latest index files to S3 and does CDN invalidations. But this is all fundamentally "eventually consistent".

S3 does have some APIs for referring to old versions of the file. So we could customize TUF to refer to the historical files already uploaded to the sparse index. But this would require a custom variation of TUF for our purposes, which does not seem to be the approach in this RFC.